federation/pkg/federatedtypes/hpa.go

```diff
@@ -0,0 +1,762 @@
+	// This is partially because a min per cluster cannot be lesser
+	// then 1, but min could be requested as 1 at federation.
+	// Additionally we first increase replicas into those clusters
+	// which already have hpa's and are in a condition to increase.
+	// This will save cluster related resources for the user, such that
+	// if an already existing cluster can satisfy users request why send
+	// the workload to another.
+	// We then go ahead to give the replicas to those which do not
+	// have any hpa. In this pass however we try to ensure that all
+	// our Max are consumed in this reconcile.
+	toDistribute.max = distributeMaxReplicas(toDistribute.max, clusterLists, rdc, currentObjs, scheduleStatus)
+
+	// We distribute min to those clusters which:
+	// 1 - can adjust min (our increase step would be only 1)
+	// 2 - which do not have this hpa and got max(increase step rdcMin)
+	// We might exhaust all min replicas here, with
+	// some clusters still needing them. We adjust this in finalise by
+	// assigning min replicas to 1 into those clusters which got max
+	// but min remains 0.
+	toDistribute.min = distributeMinReplicas(toDistribute.min, clusterLists, rdc, currentObjs, scheduleStatus)
+
+	err := a.FedUpdateStatus(fedObj, currentObjs)
+	if err != nil {
+		return nil, err
+	}
+	// Pass 3: finalise
+	return a.finaliseScheduled(fedObj, scheduleStatus)
+}
+
+func (a *HpaAdapter) FedUpdateStatus(obj pkgruntime.Object, currentObjs map[string]pkgruntime.Object) error {
+	copiedObj, err := api.Scheme.DeepCopy(obj)
+	if err != nil {
+		return fmt.Errorf("Error copying federated hpa object: %v", err)
+	}
+	if len(currentObjs) == 0 {
+		return nil
+	}
+
+	fedHpa := copiedObj.(*autoscalingv1.HorizontalPodAutoscaler)
+	gen := fedHpa.Generation
+	updatedStatus := autoscalingv1.HorizontalPodAutoscalerStatus{ObservedGeneration: &gen}
+	if fedHpa.Status.LastScaleTime != nil {
+		t := metav1.NewTime(fedHpa.Status.LastScaleTime.Time)
+		updatedStatus.LastScaleTime = &t
+	}
+	var ccup int32 = 0
+	updatedStatus.CurrentCPUUtilizationPercentage = &ccup
+
+	count := int32(0)
+	for _, lObj := range currentObjs {
+		if lObj == nil {
+			continue
+		}
+		lHpa := lObj.(*autoscalingv1.HorizontalPodAutoscaler)
+		if lHpa.Status.CurrentCPUUtilizationPercentage != nil {
+			*updatedStatus.CurrentCPUUtilizationPercentage += *lHpa.Status.CurrentCPUUtilizationPercentage
+			count++
+		}
+		if updatedStatus.LastScaleTime != nil && lHpa.Status.LastScaleTime != nil &&
+			updatedStatus.LastScaleTime.After(lHpa.Status.LastScaleTime.Time) {
+			t := metav1.NewTime(lHpa.Status.LastScaleTime.Time)
+			updatedStatus.LastScaleTime = &t
+		}
+		updatedStatus.CurrentReplicas += lHpa.Status.CurrentReplicas
+		updatedStatus.DesiredReplicas += lHpa.Status.DesiredReplicas
+	}
+
+	// Average out the available current utilisation
+	if *updatedStatus.CurrentCPUUtilizationPercentage != 0 {
+		*updatedStatus.CurrentCPUUtilizationPercentage /= count
```

Comment:
The division now is done at the end of the reconcile cycle (After modifying to conform to the schedule type adapter). Have update to weighted average correctly though. The calculation now effectively is:

*fedStatus.CurrentCpuUtilization = ((C1utilizationxC1currentReplicas) + (C2utilizationxC2currentReplicas) + ... )/(C1-currentReplicas + C2-currentReplicas + ... )*